### PR TITLE
add host & live-game-delay to user status in Chat

### PR DESCRIPTION
**Pull request by GrotheFAF**
updatePlayer:
refactor tooltip_str, icon_str
add __.png for no country2ip
remove set_color and verifySortOrder (updateUser is doing that just fine)

updateGame:
refactor gamestr, tooltip_str, icon_str
use other single sword (gold it) for host of hosted game
add title, mod, map and players info to user status tooltip (and remove url)
add playing5.png for live-game delay status icon
add liveReplayAvailable connection for updateGame
add unknown.png for user with unknown game.state (you never know...)

fix QUrl for pyqt5

### CODE REVIEW

**Comment by GrotheFAF** [comment]
The wizzards of signaling might check if I did it right.

File: src/chat/chatter.py [hunk 1]
```diff
@@ -141,6 +141,8 @@ def user_game(self, value):
 
         if self._user_game is not None:
             self._user_game.gameUpdated.connect(self.updateGame)
+            if self._user_game.state == GameState.PLAYING:
+                self._user_game.liveReplayAvailable.connect(self.updateGame)
```
Comment (Wesmania) [inline src/chat/chatter.py]: user_game is only set when a player's game changes, so if we join a game that's in another state (like OPEN), this will never get connected. We should always connect to it and check any extra conditions in the handler. Also make sure we disconnect from this signal when we change user_game again.

File: src/chat/chatter.py [hunk 2]
```diff
@@ -248,19 +250,16 @@ def updateUser(self):
...
             self.rankItem.setIcon(util.THEME.icon("chat/rank/civilian.png"))
             self.rankItem.setToolTip("IRC User")
             return
 
-        country = player.country
-        if country is not None:
-            self.setIcon(util.THEME.icon("chat/countries/%s.png" % country.lower()))
-            self.setToolTip(country)
+        country = player.country  # server sends '' for no ip2country-resolution
+        self.setIcon(util.THEME.icon("chat/countries/%s.png" % (country or '__').lower()))
```
Comment (Wesmania) [inline src/chat/chatter.py]: Changed logic, now we're checking truthiness of 'country' instead of comparing it to None. If it can be an empty string, it'd be better to check explicitly.

File: src/chat/chatter.py [hunk 3]
```diff
@@ -273,43 +272,56 @@ def updatePlayer(self):
...
+        self.rankItem.setIcon(util.THEME.icon("chat/rank/%s.png" % icon_str))
+        self.rankItem.setToolTip(tooltip_str)
 
     def updateGame(self):
         # Status icon handling
         game = self.user_game
-        player = self.user_player
         if game is not None and not game.closed():
-            url = game.url(player.id)
+            game_str = "<br/>title: " + game.title + "<br/>mod: " + game.featured_mod + "<br/>map: " + \
```
Comment (Wesmania) [inline src/chat/chatter.py]: Please use Python's format() function for this kind of stuff. This string formatting is complex enough that it deserves its own method IMO.

File: src/chat/chatter.py [hunk 4]
```diff
@@ -273,43 +272,56 @@ def updatePlayer(self):
...
+                else:
+                    icon_str, tooltip_str = "lobby", "<b>In"
             elif game.state == GameState.PLAYING:
-                self.statusItem.setIcon(util.THEME.icon("chat/status/playing.png"))
-                self.statusItem.setToolTip("Playing Game<br/>"+url.toString())
-        else:
-            self.statusItem.setIcon(QtGui.QIcon())
-            self.statusItem.setToolTip("Idle")
+                if game.launched_at:
+                    delay = time.time() - game.launched_at + 1  # + 1 async jitter
```
Comment (Wesmania) [inline src/chat/chatter.py]: We shouldn't do it like this. If we want to know precisely if the game has a live replay already, we can add extra data to Game class that gets set before it emits the live replay signal and expose it as a property.

**Review by Wesmania — CHANGES_REQUESTED** [review]

File: src/chat/chatter.py (hunk 1 above)
Comment (GrotheFAF) [reply src/chat/chatter.py]: hmm ... isn't joining a game, when it's OPEN, the usual case?
I only need that connection, if the game starts and that is a game.state change (we see chatter status change if a game starts .. so updateGame is called) ...?
So you suggest I lose the line 144?
I did a disconnect (like gameUpdated a few lines up), but that got me a lot of 'not connected' bug messages (mentioning the connect line, which step debug fine, so I don't know how to debug that and find out what's wrong. ... I guess delayed action on a connection makes it tricky)

File: src/chat/chatter.py (hunk 2 above)
Comment (GrotheFAF) [reply src/chat/chatter.py]: The None case does not happen here, but the '' does.
I thought you like it compact. You want some if ... else ... ?

File: src/chat/chatter.py (hunk 3 above)
Comment (GrotheFAF) [reply src/chat/chatter.py]: well, it grew slowly over time ... it's not that complex .. but I know you prefer some formatting ;-)
I'll see what I can do ....

File: src/chat/chatter.py (hunk 4 above)
Comment (GrotheFAF) [reply src/chat/chatter.py]: oh, you're so right ... "game.has_live_replay"    was already there, how did I miss that ...

File: src/chat/chatter.py (hunk 1 above)
Comment (GrotheFAF) [reply src/chat/chatter.py]: losing line 144, and putting the disconnect line back in done the trick. happy signaling ;-)

File: src/chat/chatter.py (hunk 2 above)
Comment (Wesmania) [reply src/chat/chatter.py]: We can do something like `country = player.country if player.country != '' else '__'`. It's not perfect, but Python doesn't have a Maybe monad just yet :/

File: src/chat/chatter.py [hunk 5]
```diff
@@ -135,12 +135,16 @@ def user_game(self):
...
         if self._user_game is not None:
             self._user_game.gameUpdated.disconnect(self.updateGame)
+            self._user_game.liveReplayAvailable.disconnect(self.updateGame)
 
         self._user_game = value
         self.updateGame()
 
         if self._user_game is not None:
             self._user_game.gameUpdated.connect(self.updateGame)
+            #game = self._user_game
```
Comment (Wesmania) [inline src/chat/chatter.py]: Don't leave commented-out code.

File: src/chat/chatter.py [hunk 6]
```diff
@@ -273,43 +274,52 @@ def updatePlayer(self):
...
         if game is not None and not game.closed():
-            url = game.url(player.id)
+            formatting = "<br/>title: {}<br/>mod: {}<br/>map: {}<br/>players: {} / {}<br/>id: {}"
+            game_str = formatting.format(game.title, game.featured_mod, maps.getDisplayName(game.mapname),
+                                         str(game.num_players), str(game.max_players), str(game.uid))
             if game.state == GameState.OPEN:
-                self.statusItem.setIcon(util.THEME.icon("chat/status/lobby.png"))
-                self.statusItem.setToolTip("In Game Lobby<br/>"+url.toString())
+                if game.password_protected:
+                    game_str = " (private) Game Lobby:</b>" + game_str
```
Comment (Wesmania) [inline src/chat/chatter.py]: You either missed a \<b\>, or it should be a \</br\>.

File: src/chat/chatter.py [hunk 7]
```diff
@@ -273,43 +274,52 @@ def updatePlayer(self):
...
+                else:
+                    icon_str, tooltip_str = "playing5", "<b>Playing:</b>" + "   LIVE DELAY (5 Min)"
+            else:  # game.state == something else
+                icon_str, tooltip_str = "unknown", "<b>Playing maybe ...</b>"
+
+        else:  # game is None or closed
+            icon_str, tooltip_str, game_str = "none", "Idle", ""
+
+        self.statusItem.setIcon(util.THEME.icon("chat/status/%s.png" % icon_str))
+        self.statusItem.setToolTip(tooltip_str + game_str)
```
Comment (Wesmania) [inline src/chat/chatter.py]: This method became a bit too convoluted IMO, it's hard to follow the logic with multiple indentations and icon + tooltip logic mixed together. We could do the icon logic separately from the tooltip logic. I also don't like how "Playing" or "Game Lobby" are duplicated - let's give it some descriptive name like "tooltip_header"and add info like 'live delay' or 'private game' separately, maybe even on a separate line.

Actually, maybe we can operate on lists of lines at first, then just join the list with `</br>`.

File: src/chat/chatter.py (hunk 2 above)
Comment (GrotheFAF) [reply src/chat/chatter.py]: did that

File: src/chat/chatter.py (hunk 5 above)
Comment (GrotheFAF) [reply src/chat/chatter.py]: sorry missed that

File: src/chat/chatter.py (hunk 6 above)
Comment (GrotheFAF) [reply src/chat/chatter.py]: nope, was ok .. but admitted complicated

File: src/chat/chatter.py (hunk 7 above)
Comment (GrotheFAF) [reply src/chat/chatter.py]: I rearanged it with header und game separate and in order

File: src/chat/chatter.py [hunk 8]
```diff
@@ -248,19 +250,16 @@ def updateUser(self):
         self._verifySortOrder()
 
     def updatePlayer(self):
-        self.set_color()
```
Comment (Wesmania) [inline src/chat/chatter.py:251]: Why was this removed?

File: src/chat/chatter.py [hunk 9]
```diff
@@ -273,43 +272,49 @@ def updatePlayer(self):
...
+                if game.has_live_replay:
+                    icon_str, tooltip_header_str = "playing", "<b>Playing" + private_str + "</b>"
+                else:
+                    icon_str, tooltip_header_str = "playing5", "<b>Playing" + private_str + "</b> - LIVE DELAY (5 Min)"
+            else:  # game.state == something else
+                icon_str, tooltip_header_str = "unknown", "<b>Playing maybe ...</b>"
+
+            formatting = "<br/>title: {}<br/>mod: {}<br/>map: {}<br/>players: {} / {}<br/>id: {}"
+            tooltip_game_str = formatting.format(game.title, game.featured_mod, maps.getDisplayName(game.mapname),
+                                                 str(game.num_players), str(game.max_players), str(game.uid))
```
Comment (Wesmania) [inline src/chat/chatter.py]: format() does conversion to string for us.

File: src/chat/chatter.py [hunk 10]
```diff
@@ -273,43 +272,49 @@ def updatePlayer(self):
...
-        self._verifySortOrder()
+            icon_str = "newplayer"
+        self.rankItem.setIcon(util.THEME.icon("chat/rank/%s.png" % icon_str))
+        self.rankItem.setToolTip(tooltip_str)
 
     def updateGame(self):
         # Status icon handling
         game = self.user_game
-        player = self.user_player
         if game is not None and not game.closed():
```
Comment (Wesmania) [inline src/chat/chatter.py]: I'd still like to see icon_str moved into a separate method. I know it seems like repeating all these cases, but it will really make things more readable (and in the future we might just want to change the tooltip separately from the icon anyway)

File: src/chat/chatter.py [hunk 11]
```diff
@@ -273,43 +272,49 @@ def updatePlayer(self):
...
             elif game.state == GameState.PLAYING:
-                self.statusItem.setIcon(util.THEME.icon("chat/status/playing.png"))
-                self.statusItem.setToolTip("Playing Game<br/>"+url.toString())
-        else:
-            self.statusItem.setIcon(QtGui.QIcon())
-            self.statusItem.setToolTip("Idle")
+                if game.has_live_replay:
+                    icon_str, tooltip_header_str = "playing", "<b>Playing" + private_str + "</b>"
+                else:
+                    icon_str, tooltip_header_str = "playing5", "<b>Playing" + private_str + "</b> - LIVE DELAY (5 Min)"
```
Comment (Wesmania) [inline src/chat/chatter.py]: Python's format() accepts keywords, and you can do stuff like "game {private}".format(private=foo, delay=bar) with no errors from the extra argument. If we set tooltip_header_str, private_str and delay string separately, then just format() the header with both, we might reduce string duplication a bit more.

File: src/chat/chatter.py [hunk 12]
```diff
@@ -432,12 +437,12 @@ def send_the_orcs():
                 menu_add("Join hosted Game", self._interactWithGame, True)
             elif game.state == GameState.PLAYING:
                 time_running = time.time() - game.launched_at
-                if time_running > 5 * 60:
+                if time_running > game.LIVE_REPLAY_DELAY_SECS:
```
Comment (Wesmania) [inline src/chat/chatter.py]: We can use game.has_live_replay here.

Comment (GrotheFAF) [reply src/chat/chatter.py]: yes we can.

File: src/chat/chatter.py (hunk 9 above)
Comment (GrotheFAF) [reply src/chat/chatter.py]: aren't we lucky ;-)

File: src/chat/chatter.py (hunk 11 above)
Comment (GrotheFAF) [reply src/chat/chatter.py]: packed it in one.

File: src/chat/chatter.py (hunk 8 above)
Comment (GrotheFAF) [reply src/chat/chatter.py:251]: not needed (see description)

File: src/chat/chatter.py [hunk 13]
```diff
@@ -273,47 +272,73 @@ def updatePlayer(self):
...
+        self.update_status_icon()
+        self.update_map()
+
+    def update_status_tooltip(self):
+        # Status tooltip handling
+        game = self.user_game
+        if game is not None and not game.closed():
+            private_str = " (private)" if game.password_protected else ""
+            if game.state == GameState.OPEN:
+                head_str = "Hosting" if game.host == self.user.name else "In Lobby"
```
Comment (Wesmania) [inline src/chat/chatter.py]: The head_str and add_str feel like something better expressed with formatting. We could do something like `header = "Hosting {private}game:"`, `header = In {private}lobby:`, `Playing{delay}:`, then just do `header.format(private=foo, delay=bar)` and use that for the second formatting.

I'm sorry for being so nitpicky, but I still have nightmares from refactoring the game tooltip string functions :( One day we're just going to use jinja2 for all this crap.

Comment (GrotheFAF) [reply src/chat/chatter.py]: come on you like it ;-)

File: src/chat/chatter.py [hunk 14]
```diff
@@ -273,47 +272,76 @@ def updatePlayer(self):
...
+
+    def update_status_tooltip(self):
+        # Status tooltip handling
+        game = self.user_game
+        if game is not None and not game.closed():
+            private_str = "(private)" if game.password_protected else ""
+            delay_str = ""
+            if game.state == GameState.OPEN:
+                if game.host == self.user.name:
+                    head_str = "Hosting {private} game</b>"
```
Comment (Wesmania) [inline src/chat/chatter.py]: Move one of spaces to private_str so we don't get double space when it's empty.

File: src/chat/chatter.py [hunk 15]
```diff
@@ -273,47 +273,76 @@ def updatePlayer(self):
...
 
         league = player.league
         if league is not None:
-            formatting = ("Division : {}\n"
-                          "Global Rating: {}")
-            self.rankItem.setToolTip(formatting.format(
-                league["division"],
-                str(int(rating))))
-            self.rankItem.setIcon(util.THEME.icon("chat/rank/%s.png" % league["league"]))
+            icon_str, tooltip_str = league["league"], "Division : " + league["division"] + "\n" + tooltip_str
```
Comment (Wesmania) [inline src/chat/chatter.py]: Was removing rating from tooltip intentional here? also format() plox :)

File: src/chat/chatter.py [hunk 16]
```diff
@@ -273,47 +273,76 @@ def updatePlayer(self):
...
         else:
-            self.rankItem.setIcon(util.THEME.icon("chat/rank/newplayer.png"))
-
-        self._verifySortOrder()
+            icon_str = "newplayer"
+        self.rankItem.setIcon(util.THEME.icon("chat/rank/%s.png" % icon_str))
```
Comment (Wesmania) [inline src/chat/chatter.py]: format(), for consistency.

**Review by Wesmania — COMMENTED** [review]
Last nitpicks, after that just make a changelog entry and I'll merge.

File: src/chat/chatter.py (hunk 15 above)
Comment (GrotheFAF) [reply src/chat/chatter.py]: guess what's in  + tooltip_str (at the end)
... and I remembered: 'format() does conversion to string for us'

**Comment by GrotheFAF** [comment]
add changelog
